zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs

```diff
@@ -0,0 +1,302 @@
+            let sum = cb.query_word();
+            let addwords_gadget = AddWordsGadget::<F, N_ADDENDS, CHECK_OVERFLOW>::construct(
+                cb,
+                addends.clone(),
+                sum.clone(),
+            );
+
+            assert_eq!(addwords_gadget.addends().len(), N_ADDENDS);
+            if !CHECK_OVERFLOW {
+                let carry_hi = addwords_gadget.carry().as_ref().unwrap();
+                cb.require_equal("carry_hi is correct", carry_hi.expr(), CARRY_HI.expr())
+            } else {
+                assert!(addwords_gadget.carry().is_none());
+            }
+
+            AddWordsTestContainer {
+                addwords_gadget,
+                addends,
+                sum,
+            }
+        }
+
+        fn assign_gadget_container(
+            &self,
+            witnesses: &[Word],
+            region: &mut CachedRegion<'_, '_, F>,
+        ) -> Result<(), Error> {
+            assert_eq!(N_ADDENDS, witnesses.len() - 1);
+            let offset = 0;
+            for (i, addend) in self.addends.iter().enumerate() {
+                let a = witnesses[i];
+                addend.assign(region, offset, Some(a.to_le_bytes()))?;
+            }
+            let sum = witnesses[N_ADDENDS];
+            self.sum.assign(region, offset, Some(sum.to_le_bytes()))?;
+
+            let addends = witnesses[0..N_ADDENDS].try_into().unwrap();
+            self.addwords_gadget.assign(region, 0, addends, sum)?;
+
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_addwords_0_0() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true>, vec![Word::from(0), Word::from(0), Word::from(0)], true);
+    }
+
+    #[test]
+    fn test_addwords_1_1() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true>, vec![Word::from(1), Word::from(1), Word::from(2)], true);
+    }
+
+    #[test]
+    fn test_addwords_1000_1000() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true>, vec![Word::from(1000), Word::from(1000), Word::from(2000)], true);
+    }
+
+    #[test]
+    fn test_addwords_to_wordmax() {
+        try_test!(AddWordsTestContainer<Fr, 2,0u64,true,>, vec![Word::MAX - 1, Word::from(1), Word::MAX],true);
+    }
+
+    #[test]
+    fn test_addwords_high_low_max() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true,>, vec![WORD_LOW_MAX, WORD_HIGH_MAX, Word::MAX],true);
+    }
+
+    #[test]
+    fn test_addwords_overflow() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true,>, [Word::MAX, Word::from(1),  Word::from(0)],false);
+    }
+
+    #[test]
+    fn test_addwords_wrong_sum0() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true,>, [Word::from(1), Word::from(0),  Word::from(0)],false);
+    }
+
+    #[test]
+    fn test_addwords_wrong_sum2() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true,>, [Word::from(2), Word::from(1),  Word::from(2)],false);
+    }
+
+    //non overflow check cases
```

Comment:
Seems like 'overflow check cases'?

Reply:
This comment is outdated as cases growing, deleted.